File: src/lib/engine/resolver.ts
/**
 * Orchestrates one full turn — port of engine/resolver.py
 *
 * Key difference: engine functions use structuredClone at top,
 * returning new state (no in-place mutation of caller's reference).
 * Resolver mutates its local clone and returns it.
 *
 * Fix: buildings are placed in the zone specified by the action (defaulting to baseZone).
 */

import { ZONES, UNITS, BUILDINGS, AGE_ADVANCE_COSTS, UPGRADES, AGE_NAMES, Zone } from "@/lib/config";
import { GameState, PlayerState, addLog, score } from "@/lib/engine/state";
import { economyTick } from "@/lib/engine/economy";
import { resolveCombat } from "@/lib/engine/combat";
import { validateAction, deductCosts, CleanAction } from "@/lib/engine/validator";
import { buildObservation } from "@/lib/prompts/builder";

export interface Agent {
  getAction(observation: object): Promise<object>;
}

export async function runTurn(
  gs: GameState,
  agentA: Agent,
  agentB: Agent,
): Promise<GameState> {
  // Work on a deep clone so callers can compare old vs new
  const state = structuredClone(gs);

  // 1. Economy tick
  economyTick(state);

  // 2. Build observations
  const obsA = buildObservation(state, "A");
  const obsB = buildObservation(state, "B");

  // 3. Parallel API calls — log errors instead of silently swallowing them
  const [rawA, rawB] = await Promise.all([
    agentA.getAction(obsA).catch((e: unknown) => {
      const msg = e instanceof Error ? e.message : String(e);
      addLog(state, `⚠ PA error: ${msg.slice(0, 120)}`);
      return {};
    }),
    agentB.getAction(obsB).catch((e: unknown) => {
      const msg = e instanceof Error ? e.message : String(e);
      addLog(state, `⚠ PB error: ${msg.slice(0, 120)}`);
      return {};
    }),
  ]);

  // 4. Validate actions
  const actionA = validateAction(rawA, state.players["A"]);
  const actionB = validateAction(rawB, state.players["B"]);

  addLog(state, `A actions: train=${JSON.stringify(actionA.train)} build=${JSON.stringify(actionA.build)} move=${JSON.stringify(actionA.move)}`);
  addLog(state, `B actions: train=${JSON.stringify(actionB.train)} build=${JSON.stringify(actionB.build)} move=${JSON.stringify(actionB.move)}`);

  // 5a. Advance age
  processAdvanceAge(state, "A", actionA);
  processAdvanceAge(state, "B", actionB);

  // 5b. Task villagers
  processTaskVillagers(state, "A", actionA);
  processTaskVillagers(state, "B", actionB);

  // 5c. Research
  processResearch(state, "A", actionA);
  processResearch(state, "B", actionB);

  // 5d. Builds
  processBuilds(state, "A", actionA);
  processBuilds(state, "B", actionB);

  // 5e. Trains (deducts costs internally)
  processTrains(state, "A", actionA);
  processTrains(state, "B", actionB);

  // 6. Deduct costs (builds only — train costs already done)
  deductCosts(actionA, state.players["A"]);
  deductCosts(actionB, state.players["B"]);

  // 7. Process moves
  processMoves(state, "A", actionA);
  processMoves(state, "B", actionB);

  // 8. Combat for each zone
  for (const zone of ZONES) {
    resolveCombat(state, zone);
  }

  // 9. Victory check
  checkVictory(state);

  state.turn += 1;
  return state;
}

function processAdvanceAge(gs: GameState, pid: "A" | "B", action: CleanAction): void {
  if (!action.advance_age) return;
  const player = gs.players[pid];
  const nextAge = player.age + 1;
  const cost = AGE_ADVANCE_COSTS[nextAge] ?? {};
  if (!canAfford(player, cost)) return;

  for (const [res, amount] of Object.entries(cost)) {
    player.resources[res] -= amount ?? 0;
    player.resourcesBanked += amount ?? 0;
  }
  player.age = nextAge as 1 | 2 | 3 | 4;
  addLog(gs, `P${pid} advanced to Age ${nextAge} (${AGE_NAMES[nextAge]})`);
}

function processTaskVillagers(gs: GameState, pid: "A" | "B", action: CleanAction): void {
  const tasks = action.task_villagers;
  if (!tasks || Object.keys(tasks).length === 0) return;
  gs.players[pid].villagerTasks = { ...tasks };
  addLog(gs, `P${pid} tasked villagers: ${JSON.stringify(tasks)}`);
}

function processResearch(gs: GameState, pid: "A" | "B", action: CleanAction): void {
  const player = gs.players[pid];
  for (const item of action.research) {
    const upgName = item.upgrade;
    if (!UPGRADES[upgName] || player.upgrades.includes(upgName)) continue;
    const upg = UPGRADES[upgName];
    const cost = upg.cost;
    if (!canAfford(player, cost)) continue;

    for (const [res, amount] of Object.entries(cost)) {
      (player.resources as Record<string, number>)[res] -= amount;
      player.resourcesBanked += amount;
    }
    player.upgrades.push(upgName);
    player.attackBonus += upg.attack_bonus;
    player.armorBonus += upg.armor_bonus;
    addLog(gs, `P${pid} researched ${upgName} (+${upg.attack_bonus} atk, +${upg.armor_bonus} armor)`);
  }
}

function processBuilds(gs: GameState, pid: "A" | "B", action: CleanAction): void {
  const player = gs.players[pid];
  for (const item of action.build) {
    const building = item.building;
    const cost = BUILDINGS[building]?.cost ?? {};
    if (!canAfford(player, cost)) continue;

    // FIX: use zone from action if provided, otherwise default to baseZone
    const zone: Zone = (item.zone as Zone) || player.baseZone;

    player.buildings[zone].push(building);

    if (building === "Wall" || building === "Tower") {
      if (!player.buildingHp[zone]) player.buildingHp[zone] = {};
      player.buildingHp[zone][building] =
        (player.buildingHp[zone][building] ?? 0) + (BUILDINGS[building].hp ?? 0);
    }

    addLog(gs, `P${pid} built ${building} in ${zone}`);
  }
}

function processTrains(gs: GameState, pid: "A" | "B", action: CleanAction): void {
  const player = gs.players[pid];
  for (const item of action.train) {
    const { unit, count } = item;
    const turns = UNITS[unit]?.train_turns ?? 1;
    const cost = UNITS[unit]?.cost ?? {};

    let queued = 0;
    for (let i = 0; i < count; i++) {
      if (!canAfford(player, cost)) break;
      for (const [res, amount] of Object.entries(cost)) {
        (player.resources as Record<string, number>)[res] -= amount;
        player.resourcesBanked += amount;
      }
      player.productionQueue.push({ unitType: unit, turnsLeft: turns });
      queued++;
    }
    if (queued > 0) addLog(gs, `P${pid} queued ${queued}×${unit} (${turns} turn(s) each)`);
  }
}

function processMoves(gs: GameState, pid: "A" | "B", action: CleanAction): void {
  const player = gs.players[pid];
  const opponent = gs.players[pid === "A" ? "B" : "A"];

  for (const item of action.move) {
    const { unit, count, from: fromZone, to: toZone } = item;

    // Cannot leave a zone with enemy units present
    const contested = Object.values(opponent.units[fromZone] ?? {}).some((c) => c > 0);
    if (contested) continue;

    const available = player.units[fromZone]?.[unit] ?? 0;
    const actual = Math.min(count, available);
    if (actual < 1) continue;

    player.units[fromZone][unit] = available - actual;
    if (!player.units[toZone]) player.units[toZone] = {};
    player.units[toZone][unit] = (player.units[toZone][unit] ?? 0) + actual;
    addLog(gs, `P${pid} moved ${actual}×${unit}: ${fromZone}→${toZone}`);
  }
}

function checkVictory(gs: GameState): void {
  const tcA = gs.players["A"].townCenterHp;
  const tcB = gs.players["B"].townCenterHp;

  if (tcA <= 0 && tcB <= 0) {
    gs.winner = "draw";
    addLog(gs, "Draw — both Town Centers destroyed!");
    return;
  }
  if (tcB <= 0) {
    gs.winner = "A";
    addLog(gs, "Player A wins — destroyed B's Town Center!");
    return;
  }
  if (tcA <= 0) {
    gs.winner = "B";
    addLog(gs, "Player B wins — destroyed A's Town Center!");
    return;
  }
  if (gs.turn >= 50) {
    const scoreA = score(gs.players["A"]);
    const scoreB = score(gs.players["B"]);
    if (scoreA > scoreB) {
      gs.winner = "A";
      addLog(gs, `Turn limit! A wins by score (${scoreA} vs ${scoreB})`);
    } else if (scoreB > scoreA) {
      gs.winner = "B";
      addLog(gs, `Turn limit! B wins by score (${scoreB} vs ${scoreA})`);
    } else {
      gs.winner = "draw";
      addLog(gs, `Turn limit! Draw (${scoreA} each)`);
    }
  }
}

function canAfford(player: PlayerState, cost: Partial<Record<string, number>>): boolean {
  return Object.entries(cost).every(([r, amount]) => (player.resources[r] ?? 0) >= (amount ?? 0));
}

File: archive/engine/resolver.py
"""
Orchestrates one full turn: economy → validate → move → combat → victory.
"""
from __future__ import annotations
import asyncio
import json
import os
from typing import Dict, Optional, Tuple

from config import TURN_LIMIT, ZONES, UNITS, BUILDINGS, AGE_ADVANCE_COSTS, UPGRADES
from engine.state import GameState, ProductionItem
from engine.economy import economy_tick
from engine.combat import resolve_combat
from engine.validator import validate_action, deduct_costs


async def run_turn(
    gs: GameState,
    agents: Dict[str, object],  # {"A": Agent, "B": Agent}
    renderer=None,
    log_dir: Optional[str] = None,
) -> Optional[str]:
    """
    Run one full game turn. Returns winner string or None if game continues.
    Mutates gs in-place.
    """
    # 1. Economy tick (uses villager_tasks from previous turn)
    economy_tick(gs)

    # 2. Build observations
    from prompts.builder import build_observation
    obs_a = build_observation(gs, "A")
    obs_b = build_observation(gs, "B")

    # 3. Parallel API calls
    raw_a, raw_b = await asyncio.gather(
        _get_agent_action(agents["A"], obs_a, "A", gs),
        _get_agent_action(agents["B"], obs_b, "B", gs),
    )

    # 4. Validate actions
    action_a = validate_action(raw_a, gs.players["A"], gs.turn)
    action_b = validate_action(raw_b, gs.players["B"], gs.turn)

    gs.add_log(f"A actions: train={action_a['train']} build={action_a['build']} move={action_a['move']}")
    gs.add_log(f"B actions: train={action_b['train']} build={action_b['build']} move={action_b['move']}")

    # 5a. Process age advances
    _process_advance_age(gs, "A", action_a)
    _process_advance_age(gs, "B", action_b)

    # 5b. Process villager tasks
    _process_task_villagers(gs, "A", action_a)
    _process_task_villagers(gs, "B", action_b)

    # 5c. Process research
    _process_research(gs, "A", action_a)
    _process_research(gs, "B", action_b)

    # 5d. Process builds
    _process_builds(gs, "A", action_a)
    _process_builds(gs, "B", action_b)

    # 5e. Process trains (deducts costs internally)
    _process_trains(gs, "A", action_a)
    _process_trains(gs, "B", action_b)

    # 6. Deduct costs (builds only — train costs already done)
    deduct_costs(action_a, gs.players["A"])
    deduct_costs(action_b, gs.players["B"])

    # 7. Process moves
    _process_moves(gs, "A", action_a)
    _process_moves(gs, "B", action_b)

    # 8. Combat resolution for each zone
    for zone in ZONES:
        resolve_combat(gs, zone)

    # 9. Victory check
    winner = _check_victory(gs)

    # 10. Render + save log
    if renderer:
        renderer.render(gs)

    if log_dir:
        _save_turn_log(gs, log_dir)

    gs.turn += 1
    return winner


async def _get_agent_action(agent, observation: dict, pid: str, gs: GameState) -> dict:
    """Call agent with timeout/error handling."""
    try:
        result = await asyncio.wait_for(
            asyncio.to_thread(agent.get_action, observation),
            timeout=60.0,
        )
        return result if isinstance(result, dict) else {}
    except asyncio.TimeoutError:
        gs.add_log(f"P{pid} agent timed out — no-op")
        return {}
    except Exception as e:
        gs.add_log(f"P{pid} agent error: {e} — no-op")
        return {}


def _process_advance_age(gs: GameState, pid: str, action: dict) -> None:
    if not action.get("advance_age"):
        return
    player = gs.players[pid]
    next_age = player.age + 1
    cost = AGE_ADVANCE_COSTS.get(next_age, {})
    if not all(player.resources.get(r, 0) >= cost.get(r, 0) for r in cost):
        return
    for res, amount in cost.items():
        player.resources[res] = player.resources.get(res, 0) - amount
        player.resources_banked += amount
    player.age = next_age
    from config import AGE_NAMES
    gs.add_log(f"P{pid} advanced to Age {next_age} ({AGE_NAMES[next_age]})")


def _process_task_villagers(gs: GameState, pid: str, action: dict) -> None:
    tasks = action.get("task_villagers", {})
    if not tasks:
        return
    player = gs.players[pid]
    player.villager_tasks = dict(tasks)
    gs.add_log(f"P{pid} tasked villagers: {tasks}")


def _process_research(gs: GameState, pid: str, action: dict) -> None:
    player = gs.players[pid]
    for item in action.get("research", []):
        upgrade_name = item.get("upgrade")
        if upgrade_name not in UPGRADES:
            continue
        if upgrade_name in player.upgrades:
            continue
        upg = UPGRADES[upgrade_name]
        cost = upg["cost"]
        if not all(player.resources.get(r, 0) >= cost.get(r, 0) for r in cost):
            continue
        for res, amount in cost.items():
            player.resources[res] = player.resources.get(res, 0) - amount
            player.resources_banked += amount
        player.upgrades.append(upgrade_name)
        player.attack_bonus += upg.get("attack_bonus", 0)
        player.armor_bonus += upg.get("armor_bonus", 0)
        gs.add_log(f"P{pid} researched {upgrade_name} (+{upg.get('attack_bonus',0)} atk, +{upg.get('armor_bonus',0)} armor)")


def _process_builds(gs: GameState, pid: str, action: dict) -> None:
    player = gs.players[pid]
    for item in action.get("build", []):
        building = item["building"]
        cost = BUILDINGS[building]["cost"]
        if not all(player.resources.get(r, 0) >= cost.get(r, 0) for r in cost):
            continue
        player.buildings[player.base_zone].append(building)
        # Track HP for Wall and Tower
        if building in ("Wall", "Tower"):
            zone_hp = player.building_hp.setdefault(player.base_zone, {})
            zone_hp[building] = zone_hp.get(building, 0) + BUILDINGS[building]["hp"]
        gs.add_log(f"P{pid} built {building} in {player.base_zone}")


def _process_trains(gs: GameState, pid: str, action: dict) -> None:
    player = gs.players[pid]
    for item in action.get("train", []):
        unit = item["unit"]
        count = item["count"]
        turns = UNITS[unit]["train_turns"]
        cost = UNITS[unit]["cost"]
        # Deduct per unit and add to queue
        for _ in range(count):
            if not all(player.resources.get(r, 0) >= cost.get(r, 0) for r in cost):
                break
            for res, amount in cost.items():
                player.resources[res] = player.resources.get(res, 0) - amount
                player.resources_banked += amount
            player.production_queue.append(ProductionItem(unit, turns))
        gs.add_log(f"P{pid} queued {count}×{unit} ({turns} turn(s) each)")


def _process_moves(gs: GameState, pid: str, action: dict) -> None:
    player = gs.players[pid]
    for item in action.get("move", []):
        unit = item["unit"]
        count = item["count"]
        from_zone = item["from"]
        to_zone = item["to"]
        available = player.units.get(from_zone, {}).get(unit, 0)
        actual = min(count, available)
        if actual < 1:
            continue
        player.units[from_zone][unit] = available - actual
        to_units = player.units.setdefault(to_zone, {})
        to_units[unit] = to_units.get(unit, 0) + actual
        gs.add_log(f"P{pid} moved {actual}×{unit}: {from_zone}→{to_zone}")


def _check_victory(gs: GameState) -> Optional[str]:
    tc_a = gs.players["A"].town_center_hp
    tc_b = gs.players["B"].town_center_hp

    if tc_a <= 0 and tc_b <= 0:
        gs.winner = "draw"
        gs.add_log("Draw — both Town Centers destroyed!")
        return "draw"
    if tc_b <= 0:
        gs.winner = "A"
        gs.add_log("Player A wins — destroyed B's Town Center!")
        return "A"
    if tc_a <= 0:
        gs.winner = "B"
        gs.add_log("Player B wins — destroyed A's Town Center!")
        return "B"

    if gs.turn >= TURN_LIMIT:
        score_a = gs.players["A"].score()
        score_b = gs.players["B"].score()
        if score_a > score_b:
            gs.winner = "A"
            msg = f"Turn limit! A wins by score ({score_a} vs {score_b})"
        elif score_b > score_a:
            gs.winner = "B"
            msg = f"Turn limit! B wins by score ({score_b} vs {score_a})"
        else:
            gs.winner = "draw"
            msg = f"Turn limit! Draw ({score_a} each)"
        gs.add_log(msg)
        return gs.winner

    return None


def _save_turn_log(gs: GameState, log_dir: str) -> None:
    os.makedirs(log_dir, exist_ok=True)
    path = os.path.join(log_dir, f"turn_{gs.turn:03d}.json")
    with open(path, "w") as f:
        json.dump(gs.to_dict(), f, indent=2)
